### ISSUE TITLE
Guix and the use of sys.argv[0]

### ISSUE THREAD

**Issue by attila-lendvai** [romanz/trezor-agent#360]
on Guix wrapper scripts are needed to set up the environment for user-facing executable programs.

`trezor-gpg init` breaks there because `libagent` uses `sys.argv[0]` to determine the `device-name`:

```
device_name = os.path.basename(sys.argv[0]).rsplit('-', 1)[0]
```

the actual backtrace:
```
Traceback (most recent call last):
  File "/gnu/store/ni8pb8iggy2jybxgcv8i0bdjnsbs0il7-trezor-agent-0.11.0/bin/.trezor-gpg-real", line 11, in <module>
    load_entry_point('trezor-agent==0.11.0', 'console_scripts', 'trezor-gpg')()
  File "/gnu/store/ni8pb8iggy2jybxgcv8i0bdjnsbs0il7-trezor-agent-0.11.0/bin/trezor_agent.py", line 6, in <lambda>
    gpg_tool = lambda: libagent.gpg.main(DeviceType)
  File "/gnu/store/k5vis47w5ns8mzz0083xms9fpv7ykcwz-python-trezor-agent-0.14.2/lib/python3.8/site-packages/libagent/gpg/__init__.py", line 320, in main
    return args.func(device_type=device_type, args=args)
  File "/gnu/store/k5vis47w5ns8mzz0083xms9fpv7ykcwz-python-trezor-agent-0.14.2/lib/python3.8/site-packages/libagent/gpg/__init__.py", line 141, in run_init
    agent_path = util.which('{}-gpg-agent'.format(device_name))
  File "/gnu/store/k5vis47w5ns8mzz0083xms9fpv7ykcwz-python-trezor-agent-0.14.2/lib/python3.8/site-packages/libagent/util.py", line 212, in wrapper
    result = func(*args, **kwargs)
  File "/gnu/store/k5vis47w5ns8mzz0083xms9fpv7ykcwz-python-trezor-agent-0.14.2/lib/python3.8/site-packages/libagent/util.py", line 248, in which
    raise OSError('Cannot find {!r} in $PATH'.format(cmd))
OSError: Cannot find '.trezor-gpg-gpg-agent' in $PATH
```

if there is another simple and more resilient way to implement this, then i'd welcome that, because resolving this on the Guix side is not trivial. if there isn't, then feel free to close this.

FTR, the Guix side of the issue: https://lists.gnu.org/archive/html/guix-devel/2017-11/msg00017.html

**Comment by romanz** [comment]
Many thanks for reporting this issue and apologies for the (very) delayed response!
Please try https://github.com/romanz/trezor-agent/pull/368 :)

**Comment by attila-lendvai** [comment]
can't jump into it right now, but it should work.

FTR, the Guix side of the discussion (for now including a patch that this fix obsoletes): http://issues.guix.gnu.org/50751

**Comment by romanz** [comment]
The PR is merged, will be happy to tag a new version soon :)

**Comment by attila-lendvai** [comment]
great, thanks!

i ran a quick check: built the guix package pointing to this commit (and of course without my patch for the argv stuff), and a quick `trezor-gpg init` ran as expected.

**Comment by romanz** [comment]
Cool, many thanks!

**Comment by romanz** [comment]
`0.14.4` is tagged and released:
https://github.com/romanz/trezor-agent/releases/tag/v0.14.4
https://pypi.org/project/libagent/0.14.4

**Comment by attila-lendvai** [comment]
@romanz just a note, so that you're aware of it:

TL;DR: `pinentry-gtk-2` might be broken on Guix.

on Guix, the importing of the public key inside `trezor-gpg init` hangs, unless i override the passphrase pinentry from the default `pinentry` (which is a symlink to `pinentry-gtk-2`), to `pinentry-gnome3` (any running agent must be killed for this to have effects).

`strace` shows that the hang is when communicating with the agent, asking for a passphrase, but it hangs with nothing happening onscreen.

i don't know why the/a trezor passphrase is asked for when importing the public key, but that could be just my lack of understanding.

FTR,

```
trezor-gpg init -v -v -v --passphrase-entry-binary /home/alendvai/.guix-profile/bin/pinentry-gnome3 --pin-entry-binary /home/alendvai/.guix-profile/bin/trezor-gpg-pinentry-tk "Attila Lendvai <attila@lendvai.name>"
```

and this is what i copied from the logs to test with `strace`:

```
strace /gnu/store/cc0r4spdjr25pisaj4fs7pcyb62v1r2x-gnupg-2.2.29/bin/gpg --homedir /home/alendvai/.gnupg/trezor -vvv --import /home/alendvai/.gnupg/trezor/pubkey.asc
```

**Comment by attila-lendvai** [comment]
@romanz 

strangely enough, when i kill the agent, and run the gpg import line, then it works without asking for any passphrase when trezor is unplugged:

```
$ rm -f ~/.gnupg/trezor/trustdb.gpg* ~/.gnupg/trezor/pubring.kbx*
alendvai@lelap ~/workspace/guix/nonguix$ ps afxu | grep agent
alendvai 28917  0.0  0.0   6392  1800 pts/1    S+   14:29   0:00      \_ grep --color=auto agent
alendvai@lelap ~/workspace/guix/nonguix$ /gnu/store/cc0r4spdjr25pisaj4fs7pcyb62v1r2x-gnupg-2.2.29/bin/gpg --homedir /home/alendvai/.gnupg/trezor --import /home/alendvai/.gnupg/trezor/pubkey.asc
gpg: keybox '/home/alendvai/.gnupg/trezor/pubring.kbx' created
gpg: /home/alendvai/.gnupg/trezor/trustdb.gpg: trustdb created
gpg: key 1D176925DC333995: public key "Attila Lendvai <attila@lendvai.name>" imported
gpg: Total number processed: 1
gpg:               imported: 1
alendvai@lelap ~/workspace/guix/nonguix$ ps afxu | grep agent
alendvai 28983  0.0  0.0   6392  1764 pts/1    S+   14:29   0:00      \_ grep --color=auto agent
alendvai 28960  2.7  0.1  45776 29604 ?        Sl   14:29   0:00 /gnu/store/nckjv3ccwdi6096j478gvns43ssbls2p-python-wrapper-3.8.2/bin/python /gnu/store/5n0rx6v1vq298apd7ppj607s9hicpc48-trezor-agent-0.11.0-1/bin/.trezor-gpg-agent-real -vv --pin-entry-binary=/home/alendvai/.guix-profile/bin/trezor-gpg-pinentry-tk --passphrase-entry-binary=/home/alendvai/.guix-profile/bin/pinentry-gtk-2 --cache-expiry-seconds=inf --homedir /home/alendvai/.gnupg/trezor --use-standard-socket --daemon
```

but when the trezor is plugged in, then the pin is asked for (i.e. not the passphrase), trezor gets opened, and it hangs after the import, when it's asking for the passphrase:

```
$ /gnu/store/cc0r4spdjr25pisaj4fs7pcyb62v1r2x-gnupg-2.2.29/bin/gpg --homedir /home/alendvai/.gnupg/trezor --import /home/alendvai/.gnupg/trezor/pubkey.asc
gpg: keybox '/home/alendvai/.gnupg/trezor/pubring.kbx' created
gpg: /home/alendvai/.gnupg/trezor/trustdb.gpg: trustdb created
gpg: key [id]: public key "Attila Lendvai <foo@example.com>" imported
```

it works when i change the pinentry to `pinentry-gnome3`, but the question remains: why is trezor opened when importing the public key, and the passphrase asked for?

i think this is probably related to trezor-agent, too.

**Comment by romanz** [comment]
> why is trezor opened when importing the public key, and the passphrase asked for?

TREZOR allows the user to enable [passphrase-based key derivation](https://wiki.trezor.io/Passphrase).
If you don't use this feature, you can disable it on your device using:
```
$ trezorctl set passphrase --help
Usage: trezorctl set passphrase [OPTIONS] COMMAND [ARGS]...

  Enable, disable or configure passphrase protection.

Options:
  --help  Show this message and exit.

Commands:
  disabled  Disable passphrase.
  enabled   Enable passphrase.

```

**Comment by attila-lendvai** [comment]
i understand that (i think), but when gpg is **importing** the public key?

```
gpg --homedir ~/.gnupg/trezor --import ~/.gnupg/trezor/pubkey
```

also note that this gpg import command runs fine (doesn't hang) when the trezor device is entirely unplugged.

**Comment by romanz** [comment]
Could you please share the `~/.gnupg/trezor/gpg-agent.log` when the import gets stuck?
